flocker/node/agents/testtools.py

```diff
@@ -58,3 +96,156 @@ def setUp(self):
+    Interface for cleanup operations.
+    """
+
+    def execute(blockdevice_manager):
+        """
+        Perform the cleanup operation.
+
+        :param blockdevice_manager: The :class:`IBlockDeviceManager` provider
+            to use to execute the cleanup.
+
+        :raises CleanupError: If an error occurs that does not indicate a bug
+            in the code and should not stop cleanup execution.
+        """
+
+
+@implementer(_ICleanupOperation)
+class _UnmountCleanup(PClass):
+    """
+    Object for cleanup by unmounting.
+
+    :ivar FilePath path: The path to unmount.
+    """
+    path = field(type=FilePath)
+
+    def execute(self, blockdevice_manager):
+        try:
+            blockdevice_manager.unmount(self.path)
+        except UnmountError as e:
+            raise CleanupError(error=e)
+
+
+class CleanupBlockDeviceManager(proxyForInterface(IBlockDeviceManager)):
+    """
+    Proxies to another :class:`IBlockDeviceManager` provider, and records every
+    created mount, symlink, etc. for cleanup later.
+
+    This is a test helper class for tests that use
+    :class:`IBlockDeviceManager`, and don't want to manually manage cleanup of
+    all of the mounts and symlinks.
+
+    Note: This does not behave precisely correct for mounted blockdevices that
+    are unmounted by mount point.
+
+    :ivar _cleanup_operations: A list of operations to perform upon cleanup in
+        reverse order. These must provide :class:`_ICleanupOperation`.
+    """
+    def __init__(self, original):
+        super(CleanupBlockDeviceManager, self).__init__(original)
+        self._cleanup_operations = []
+
+    def mount(self, blockdevice, mountpoint):
+        self._cleanup_operations.append(_UnmountCleanup(path=blockdevice))
+        return self.original.mount(blockdevice, mountpoint)
+
+    def unmount(self, unmount_path):
+        unmount_index = next(iter(
+            -index
+            for index, op in enumerate(reversed(self._cleanup_operations), 1)
+            if op == _UnmountCleanup(path=unmount_path)
+        ), None)
+        if unmount_path is not None:
+            self._cleanup_operations.pop(unmount_index)
+        return self.original.unmount(unmount_path)
+
+    def make_tmpfs_mount(self, mountpoint):
+        self._cleanup_operations.append(_UnmountCleanup(path=mountpoint))
+        return self.original.make_tmpfs_mount(mountpoint)
+
+    def bind_mount(self, source_path, mountpoint):
+        self._cleanup_operations.append(_UnmountCleanup(path=mountpoint))
+        return self.original.bind_mount(source_path, mountpoint)
+
+    def cleanup(self):
+        """
+        Perform all cleanup operations.
+        """
+        cleanup_errors = []
+        for operation in reversed(self._cleanup_operations):
+            try:
+                operation.execute(self.original)
+            except CleanupError as e:
+                cleanup_errors.append(e)
+        if cleanup_errors:
+            raise cleanup_errors[0].error
```

Comment:
That's pretty cool, but unfortunately I want to be able to remove cleanup operations, and it doesn't look like fixtures help with that very much? Let me know if I'm reading that wrong

https://github.com/testing-cabal/fixtures/blob/master/fixtures/fixture.py#L77 seems to suggest they don't.

I'll fix up `blockdevice_manager_for_test` to return a fixture rather than take a test case though. I think that makes sense.